### PR TITLE
Add support for Arrow reads

### PR DESCRIPTION
**Pull request by jphalip**

### CODE REVIEW

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/utils/arrow/ArrowSerializer.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/utils/arrow/ArrowSerializer.java:70]: Not relevant. Works as intended.

**Review by VinayLondhe14 — APPROVED** [review]
LGTM but can we split the integration tests into multiple classes instead of creating them all in one class?

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java
Comment (davidrabinowitz) [inline connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java:221]: No need to create the injector just to create `BigQueryClientFactory`. It is serializable and can be passed as a field of the split.

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/arrow/BigQueryArrowInputFormat.java
Comment (davidrabinowitz) [inline connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/arrow/BigQueryArrowInputFormat.java:40]: Any reason to delegate to a static method in the split? it seems that the code can be here

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (davidrabinowitz) [inline connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:173]: Please Goole Truth assert as they are more readable `assertThat(rows).isEmpty();`

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (davidrabinowitz) [inline connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:177]: Please split to two methods - testReadEmptyTable_mr(), testReadEmptyTable_tez(), so when there is a failure we will know if it's in a single engine or both (and same for other relevant methods)

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (davidrabinowitz) [inline connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:164]: Notice that tearDown() won't be called on assertion failure. It is best to mark setup() and tearDown() with @Before and @After, and use the @RunWith(Parameterized.class) to provide the engines.

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (davidrabinowitz) [inline connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:291]: no asserts in this tests

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/BigQueryStorageHandler.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/BigQueryStorageHandler.java:43]: I see what you mean. `BigQueryArrowInputFormat` is only referred to in `BigQueryStorageHandler.java` though so I'm not sure it's worth moving it to HiveBigQueryConfig. Would it be ok to just change the logic here to this?

```
    Injector injector =
        Guice.createInjector(
            new HiveBigQueryConnectorModule(conf, null));
    HiveBigQueryConfig opts = injector.getInstance(HiveBigQueryConfig.class);
    DataFormat readDataFormat = opts.getReadDataFormat();
    if (readDataFormat.equals(DataFormat.ARROW)) {
      return BigQueryArrowInputFormat.class;
    } else {
      throw new RuntimeException("Invalid readDataFormat: " + readDataFormat);
    }
```

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/arrow/BigQueryArrowInputFormat.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/arrow/BigQueryArrowInputFormat.java:40]: Yes, it's because the same code is needed for avro reads, which comes in the next PR.

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (jphalip) [reply connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:177]: So that's what I had originally but that exploded into a huge number of methods. That's because it's useful to test for multiple execution engine, multiple read data format, and multiple write method. The number of methods got a bit unwieldy, which is why I used nested loops. I wonder if maybe there's a way to clearly display the loop values when there's a failure?

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (jphalip) [reply connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:164]: Yes good point. I'm planning to refactor everything using a TestResource. Would it be ok if I do that in the end after all the PRs have been merged? Otherwise it's going to get difficult rebasing every single PRs as that's going to create a lot of conflicts.

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (jphalip) [reply connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:291]: Yeah, this is more of a smoke test. Do you think we absolutely need an assertion?

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java:115]: Actually, in a later PR I remove those fields, so I suggest we don't add them here since we'd have to remove them anyways.

File: connector/pom.xml
Comment (jphalip) [reply connector/pom.xml:110]: @davidrabinowitz Looks like the Spark connector also uses that same version of Truth. What do you advise?

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java:208]: Are there benefits in passing it as a serialized field?

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryConstantDesc.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryConstantDesc.java:64]: Good question. I should probably at least add a test to check.

Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryConstantDesc.java:64]: Also, just so it doesn't hold up this PR, could we maybe keep it as a future feature request?

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (jphalip) [reply connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:292]: As a smoke test, it just checks that no exception gets thrown. I have other tests that check for validity of some types of filters. More tests could be added to verify the validity of all types of filters. But I thought a smoke test would be good enough for now.

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java
Comment (jphalip) [reply connector/src/test/java/com/google/cloud/hive/bigquery/connector/IntegrationTests.java:178]: Ok, I see there's also a `@ParameterizedTest` decorator combined with `@ValueSource`. Would that work?

Also, would it be ok to change that once all the PRs have been merged to avoid too many rebase conflicts?

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java:208]: Ok, though we would still use the injector to get the `HiveBigQueryConfig`, right?

Also, to pass the serialized version of `BigQueryClientFactory`, you would use `writeObject` and `readObject`, right?

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/config/HiveBigQueryConfig.java
Comment (davidrabinowitz) [inline connector/src/main/java/com/google/cloud/hive/bigquery/connector/config/HiveBigQueryConfig.java:86]: The member itself should be a regular TableID, not optional

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/config/HiveBigQueryConfig.java
Comment (davidrabinowitz) [inline connector/src/main/java/com/google/cloud/hive/bigquery/connector/config/HiveBigQueryConfig.java:138]:
```
config.tableId = tableId.orElse(TableId.of(
                  conf.get(HiveBigQueryConfig.PROJECT_KEY),
                  conf.get(HiveBigQueryConfig.DATASET_KEY),
                  conf.get(HiveBigQueryConfig.TABLE_KEY)));
```

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java
Comment (davidrabinowitz) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java:212]: You can pass also the relevant configuration as well, it is serializable for that reason. `Serializable` classes do not need `writeObject` and `readObject` methods, then can have them for customizing the write (but usually are redundant)

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java:98]: @davidrabinowitz Is it ok to ignore?

File: connector/src/main/java/com/google/cloud/hive/bigquery/connector/config/HiveBigQueryConfig.java
Comment (jphalip) [reply connector/src/main/java/com/google/cloud/hive/bigquery/connector/config/HiveBigQueryConfig.java:138]: I've changed the code here slightly, so I'm not sure this format would apply any more. It will be replaced in a later PR anyways.

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**
